# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmTable.cpp
# Commit: 1a73cfde72e8
# Full Hash: 1a73cfde72e8f4d19bdcfdbffa9fbc78bba01e24
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2021-09-28 09:54:33
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Introduce indirect stubs to optimize call_indirect. r=lth
#   
#   This patch introduces indirect stubs.
#   An indirect stub is a stub that takes care of any switching activities needed for call_indirect.
#   Before this patch, we have to conservatively assume that any call_indirect's target can be from a foreign instance,
# ==============================================================================

diff -r da3c5d8492db -r 1a73cfde72e8 js/src/wasm/WasmTable.cpp
--- a/js/src/wasm/WasmTable.cpp	Tue Sep 28 05:16:23 2021 +0000
+++ b/js/src/wasm/WasmTable.cpp	Tue Sep 28 05:24:15 2021 +0000
@@ -38,6 +38,7 @@
       functions_(std::move(functions)),
       elemType_(desc.elemType),
       isAsmJS_(desc.isAsmJS),
+      importedOrExported(desc.importedOrExported),
       length_(desc.initialLength),
       maximum_(desc.maximumLength) {
   MOZ_ASSERT(repr() == TableRepr::Func);
@@ -50,6 +51,7 @@
       objects_(std::move(objects)),
       elemType_(desc.elemType),
       isAsmJS_(desc.isAsmJS),
+      importedOrExported(desc.importedOrExported),
       length_(desc.initialLength),
       maximum_(desc.maximumLength) {
   MOZ_ASSERT(repr() == TableRepr::Ref);
@@ -151,17 +153,38 @@
   MOZ_ASSERT(isFunction());
 
   const FunctionTableElem& elem = getFuncRef(index);
-  if (!elem.code) {
+  if (!elem.tls) {
     fun.set(nullptr);
     return true;
   }
 
   Instance& instance = *elem.tls->instance;
-  const CodeRange& codeRange = *instance.code().lookupFuncRange(elem.code);
+  const CodeRange* codeRange =
+      instance.code().lookupIndirectStubRange(elem.code);
+  if (!codeRange) {
+    codeRange = instance.code().lookupFuncRange(elem.code);
+  }
+  MOZ_ASSERT(codeRange);
 
-  RootedWasmInstanceObject instanceObj(cx, instance.object());
-  return instanceObj->getExportedFunction(cx, instanceObj,
-                                          codeRange.funcIndex(), fun);
+  // If the element is a wasm function imported from another
+  // instance then to preserve the === function identity required by
+  // the JS embedding spec, we must set the element to the
+  // imported function's underlying CodeRange.funcCheckedCallEntry and
+  // Instance so that future Table.get()s produce the same
+  // function object as was imported.
+  JSFunction* callee = nullptr;
+  Instance* calleeInstance =
+      instance.getOriginalInstanceAndFunction(codeRange->funcIndex(), &callee);
+  RootedWasmInstanceObject calleeInstanceObj(cx, calleeInstance->object());
+  uint32_t calleeFunctionIndex = codeRange->funcIndex();
+  if (callee && (calleeInstance != &instance)) {
+    const Tier calleeTier = calleeInstance->code().bestTier();
+    calleeFunctionIndex =
+        calleeInstanceObj->getExportedFunctionCodeRange(callee, calleeTier)
+            .funcIndex();
+  }
+  return WasmInstanceObject::getExportedFunction(cx, calleeInstanceObj,
+                                                 calleeFunctionIndex, fun);
 }
 
 void Table::setFuncRef(uint32_t index, void* code, const Instance* instance) {
@@ -183,7 +206,7 @@
   }
 }
 
-void Table::fillFuncRef(uint32_t index, uint32_t fillCount, FuncRef ref,
+bool Table::fillFuncRef(uint32_t index, uint32_t fillCount, FuncRef ref,
                         JSContext* cx) {
   MOZ_ASSERT(isFunction());
 
@@ -191,7 +214,7 @@
     for (uint32_t i = index, end = index + fillCount; i != end; i++) {
       setNull(i);
     }
-    return;
+    return true;
   }
 
   RootedFunction fun(cx, ref.asJSFunction());
@@ -208,14 +231,16 @@
 #endif
 
   Instance& instance = instanceObj->instance();
-  Tier tier = instance.code().bestTier();
-  const MetadataTier& metadata = instance.metadata(tier);
-  const CodeRange& codeRange =
-      metadata.codeRange(metadata.lookupFuncExport(funcIndex));
-  void* code = instance.codeBase(tier) + codeRange.funcCheckedCallEntry();
+  void* code = instance.createIndirectStub(funcIndex);
+  if (!code) {
+    return false;
+  }
+
   for (uint32_t i = index, end = index + fillCount; i != end; i++) {
     setFuncRef(i, code, &instance);
   }
+
+  return true;
 }
 
 AnyRef Table::getAnyRef(uint32_t index) const {